### PR TITLE
Add API for wire instructions

### PR DESCRIPTION
**Pull request by aryanbhasin**

### CODE REVIEW

File: lib/circlex/struct/wire_instructions.ex [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        %__MODULE__{
+            beneficiary: fetch(wire_instructions, :beneficiary),
+            beneficiary_bank: fetch(wire_instructions, :beneficiaryBank),
+            tracking_ref: fetch(wire_instructions, :trackingRef),
+            virtual_account_enabled: fetch(wire_instructions, :virtualAccountEnabled)
+        }
+    end
+
+    # Accepts Struct.BankAccount, since most wire instructions are hard-coded
+    def serialize(wire_instructions) do
```
Comment (aryanbhasin) [inline lib/circlex/struct/wire_instructions.ex]: could improve naming of argument -- if a `bank_account` is passed in, we should not call it `wire_instructions`

File: lib/circlex/emulator/api/core/bank_accounts_api.ex [hunk 2]
```diff
@@ -22,6 +23,14 @@ defmodule Circlex.Emulator.Api.Core.BankAccountsApi do
     end
   end
 
+  # https://developers.circle.com/reference/payments-bank-accounts-wires-get-id-instructions
+  @route "/wires/:bank_account_id/instructions"
+  def get_wire_instructions(%{bank_account_id: bank_account_id}) do
+    with {:ok, bank_account_sending_wire} <- BankAccountState.get_bank_account(bank_account_id) do
+      {:ok, WireInstructions.serialize(bank_account_sending_wire)}
```
Comment (aryanbhasin) [inline lib/circlex/emulator/api/core/bank_accounts_api.ex]: change variable name for `bank_account_sending_wire`

**Review by aryanbhasin — COMMENTED** [review]
`WireInstructions.serialize(argument)` needs a better argument name

File: lib/circlex/struct/wire_instructions.ex (hunk 1 above)
Comment (hayesgm) [reply lib/circlex/struct/wire_instructions.ex]: yeah, can you pattern match on the struct to ensure it is what you think it is? As in

```diff
-def serialize(wire_instructions) do
+def serialize(bank_account=%Circlex.Struct.BankAccount{}) do
```

But overall, I'm a little confused by the exact overlap of "BankAccount" and "WireInstructions"?

File: lib/circlex/struct/wire_instructions.ex [hunk 3]
```diff
@@ -0,0 +1,43 @@
+defmodule Circlex.Struct.WireInstructions do
```
Comment (hayesgm) [inline lib/circlex/struct/wire_instructions.ex:1]: Can we also add struct test for this module?

File: test/circlex/emulator/api/core/bank_accounts_api_test.exs [hunk 4]
```diff
@@ -97,10 +99,38 @@ defmodule Circlex.Emulator.Api.Core.BankAccountsApiTest do
...
+                "city": "NEW YORK",
+                "country": "US",
+                "currency": "USD",
+                "name": "CRYPTO BANK",
+                "postalCode": "1001",
+                "routingNumber": "999999999",
+                "swiftCode": "CRYPTO99"
+              },
+              trackingRef: "CIR3KX3L99",
+              virtualAccountEnabled: true
```
Comment (hayesgm) [inline test/circlex/emulator/api/core/bank_accounts_api_test.exs:130]: is there a corner-case where virtualAccountEnabled = false that would make a sensible second test? Could be no.

File: lib/circlex/emulator/state/bank_account_state.ex [hunk 5]
```diff
@@ -48,7 +48,8 @@ defmodule Circlex.Emulator.State.BankAccountState do
        billing_details: billing_details,
        bank_address: bank_address,
        create_date: State.next(:date),
-       update_date: State.next(:date)
+       update_date: State.next(:date),
+       virtual_account_number: State.next(:virtual_account_number)
```
Comment (hayesgm) [inline lib/circlex/emulator/state/bank_account_state.ex:52]: Do we _always_ generate a virtual account number or is this some kind of opt-in thing?

**Review by hayesgm — COMMENTED** [review]
Looks great! Minor comments around some tests, but I think this is awesome.

File: lib/circlex/emulator/state.ex [hunk 6]
```diff
@@ -243,6 +243,9 @@ defmodule Circlex.Emulator.State do
   defp generate_type(:external_ref),
     do: "EXTREF" <> to_string(System.unique_integer([:positive]))
 
+  defp generate_type(:virtual_account_number),
+    do: Enum.random(100_000_000_000..999_999_999_999) |> to_string()
```
Comment (cwang25) [inline lib/circlex/emulator/state.ex:247]: Just curious, so it will always generate a different virtual_account_number every time ?
So if from treasury app, we called get_wire_instruction with same bank id, will we get two different virtual account number or two identical account number ?

Comment (cwang25) [reply lib/circlex/emulator/state.ex:247]: Oh I see. only create when creating bank info. When instruction api is invoke it only retrieve the data.

👍  :P

File: lib/circlex/emulator/api/core/bank_accounts_api.ex (hunk 2 above)
Comment (coburncoburn) [inline lib/circlex/emulator/api/core/bank_accounts_api.ex]: oh uh, this isn't really right. The instructions are always going to be Circles bank, with a dynamic Account Number. The BankAccount here is the customer's bank

File: lib/circlex/struct/wire_instructions.ex (hunk 1 above)
Comment (coburncoburn) [reply lib/circlex/struct/wire_instructions.ex]: bank account and wire instructions are different objects

File: lib/circlex/emulator/api/core/bank_accounts_api.ex (hunk 2 above)
Comment (aryanbhasin) [reply lib/circlex/emulator/api/core/bank_accounts_api.ex]: yup, but the [API](https://developers.circle.com/reference/payments-bank-accounts-wires-get-id-instructions) still takes in the customer's bank account to return the virtual account number corresponding to that account.

what I'm actually doing here is taking in the VAN and trackingRef stored in `Struct.BankAccount` to serialize a `WireInstructions` struct like this--
Screen Shot 2022-08-12 at 3 56 23 PM

Comment (coburncoburn) [reply lib/circlex/emulator/api/core/bank_accounts_api.ex]: okay, so it takes the argument as the bank account, and returns something of different shape. I think most of these are serialize functions are returning an object of the same shape, just going from map -> struct or struct -> map

Comment (coburncoburn) [reply lib/circlex/emulator/api/core/bank_accounts_api.ex]: I guess the interface is "thing with virtual account number and tracking ref keys" ? That seems okay.

File: lib/circlex/struct/wire_instructions.ex (hunk 3 above)
Comment (aryanbhasin) [reply lib/circlex/struct/wire_instructions.ex:1]: [fixed](https://github.com/compound-finance/circlex/pull/3/commits/90710b1ca6d4b81387c2f5692a570d5fc3ae3356)

**Review by hayesgm — APPROVED** [review]
Looks perfect. Can you bump the version in `mix.exs` and `README.md` and note in `README.md` if "Client Features" or "Emulator Features" needs any additional checkmarks?

**Review by hayesgm — APPROVED** [review]
Ship it.
